include/audio_backend/__asio_audio_device.h

```diff
@@ -0,0 +1,591 @@
+  vector<buffer_size_t> _buffer_sizes;
+  buffer_size_t _buffer_size = 0;
+  bool _running = false;
+
+  ASIOSampleType _sample_type;
+  ASIOSamples _sample_position = 0;
+
+  using __asio_callback_t = function<void(audio_device&, audio_device_io<float>&)>;
+  __asio_callback_t _user_callback;
+  audio_device_io<__asio_common_sample_type> _io;
+  vector<float> _input_samples;
+  vector<float> _output_samples;
+
+  using fill_buffers = function<void(long index)>;
+  fill_buffers _read;
+  fill_buffers _write;
+
+  vector<ASIOBufferInfo> _asio_buffers;
+  ASIOCallbacks _asio_callbacks;
+
+  static audio_device* instance(audio_device* d = nullptr) {
+    static audio_device* device = d;
+    return device;
+  }
+};
+
+class audio_device_list : public forward_list<audio_device> {
+};
+
+class __reg_key_reader final
+{
+public:
+  __reg_key_reader(HKEY key, const string& subkey) {
+    const auto result = RegOpenKeyExA(key, subkey.c_str(), 0, KEY_READ, &_key);
+
+    if (ERROR_SUCCESS != result) {
+      throw runtime_error("Failed to read info from registry: 0x" + result);
+    }
+  }
+
+  ~__reg_key_reader() {
+    if (_key) {
+      RegCloseKey(_key);
+    }
+  }
+
+  vector<string> subkeys() const {
+    constexpr int max_asio_name_length{32};
+    char name[max_asio_name_length];
+    vector<string> keys;
+
+    for (DWORD index = 0; ; ++index) {
+      DWORD size{max_asio_name_length};
+      const auto result = RegEnumKeyEx(_key, index, name, &size, nullptr, nullptr, nullptr, nullptr);
+
+      if (result != ERROR_SUCCESS) {
+        break;
+      }
+
+      keys.emplace_back(name);
+    }
+
+    return keys;
+  }
+
+  __reg_key_reader subkey(const string& name) const {
+    return {_key, name};
+  }
+
+  string value(const char* name) const {
+    DWORD size{0};
+    RegQueryValueExA(_key, name, nullptr, nullptr, nullptr, &size);
+
+    string value(size + 1, 0);
+    RegQueryValueExA(_key, name, nullptr, nullptr, LPBYTE(value.data()), &size);
+
+    return value;
+  }
+
+private:
+  HKEY _key{0};
+};
+
+class __asio_devices {
+public:
+  static __asio_devices& get() {
+    static __asio_devices instance;
+    return instance;
+  }
+
+  ~__asio_devices() {
+    CoUninitialize();
+  }
+
+  optional<audio_device> default_input() const {
+    auto devices = inputs();
+    if (devices.empty()) {
+      return {};
+    }
+    return *devices.begin();
+  }
+
+  optional<audio_device> default_output() const {
+    auto devices = outputs();
+    if (devices.empty()) {
+      return {};
+    }
+    return *devices.begin();
+  }
+
+  audio_device_list inputs() const {
+    return enumerate(audio_direction::in);
+  }
+
+  audio_device_list outputs() const {
+    return enumerate(audio_direction::out);
+  }
+
+private:
+  __asio_devices() {
+    CoInitialize(nullptr);
+  }
+
+  audio_device_list enumerate(audio_direction direction) const {
+    audio_device_list devices;
+
+    __reg_key_reader asio_reg(HKEY_LOCAL_MACHINE, "software\\asio");
```

Comment:
Prefer a raw string literal here to avoid having to escape the backslash.

Reply:
Fixed in 87dc8b7